### DISCUSSION TITLE
[C128] How to shift 2 bytes to compensate for the .prg load address

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#2130] [Q&A]
Hi Community, 

I am going around in circles a bit so decided to ask your guidance as I think the answer is rather simple but I keep missing it. 

The thing is for my project I generate 4 memory files "ram1.prg", "ram2.prg" and "ram3.prg".
So I can BLOAD them into the respective memory banks of the C128. 
Per example BLOAD "ram2.prg",b2,p8192

However, as these are binary files and the C128 automatic adds the starting address to the .prg my code is shifted.
Which is a known challenge so what I have to do manual is; 

Compiled Code File; 
![image](https://github.com/cc65/cc65/assets/66948775/58f535d1-79b6-4e78-9a1b-a129b75253aa)

Manual 'Add' two 00 00
![image](https://github.com/cc65/cc65/assets/66948775/25d3e643-379e-4819-9faf-15551b9a65e2)

So when I put it on the floppy and load the binary.
The starting address will be go into those two 00 00.
It will be ignored during my execution and all works fine. 

If I don't do this. The first two bytes of the actual file will be overwritten with the load address. 
Then my code is 'corrupted'  

However, I want this manual effort to be avoided. 

Now when it would be a piece of .asm code. 
I know the trick to do the ADDR+2. 
But that doesn't work in my case as it comes from compiled C code with the CC65.

Furthermore, if I do this, via a work around, all my code is pushed forward two places.
Which makes that the jumps don't work anymore as the addresses changed.  

Then the other trick I tried is what is used with the OVERLAY Demo program 
Which works like a charm when I use it to compile the OVERLAY Demo. 

![image](https://github.com/cc65/cc65/assets/66948775/299f7c16-adda-4707-914e-5f7c8d9c6986)

But somehow when I try that in my own environment; 
 
![image](https://github.com/cc65/cc65/assets/66948775/d10b2617-d227-4129-ac79-b566a5564091)

![image](https://github.com/cc65/cc65/assets/66948775/d1e74215-d29e-42c4-9882-e0b1d8f196c9)

I get an error message:
![image](https://github.com/cc65/cc65/assets/66948775/21baa7a8-1ad5-45b1-9468-27d89208ab57)

**In short the question;** 

I am creating an C program with different pieces of code going into different memory files. 

What do I have to add where to make sure that; 

1 - The generated memory file has this compensation for the load address.
So when it is BLOAD 
It starts at the address it should be (P8192 in the above) beginning executing the actual code generated by the compiler

2 - It doesn't 'move' my code addresses so the jumps etc. don't work anymore. 

3 - I don't get the error code segment does not exist (or other errors) 

I understand the above is rather descriptive but I hope somebody sees the obvious flaw and can guide me in the right direction.

Thanks in advance!

**Comment by mrdudz** [comment]
It tells you the segment does not exist, because you didnt emit data to it. For your case there are two possible solutions:
```
.segment "LOADADDRRAM2"
. word $1234 ; whatever load address you want
```
OR, if you dont care about the address, then you could add ```fill=yes``` to the segment definition in the linker config. Actually, you can do both :)

**Reply by RetrofanC** [reply]
MrDudz, thanks again much appreciated I got it to work in .asm!! (-8 

I am looking for a way to make it work from .c as well. 

As I understand I cannot use ".word" kind of pseudo instructions as inline asm right? 
_(If there is any work around which still makes it possible I would be curious)_ 

So I decided that I try this, but this gives an error; 
#pragma code-name (push, "LOADADDRRAM2") 
__asm__ ("nop"); 
__asm__ ("nop"); 
#pragma code-name (pop)


Error: 
__asm__ is not allowed here
__asm__ is not allowed here

So I made it this instead; 

#pragma code-name (push, "LOADADDRRAM2") 
void add_two_bytes(void){
__asm__ ("nop"); 
__asm__ ("nop");}
 #pragma code-name (pop)

If there is any more clever way I am all ears (-' 

But for now I will close this as answered!

**Comment by mrdudz** [comment]
declare a static array!

#pragma code-name (push, "LOADADDRRAM2")
unsigned char foo[2] = { 0x01, 0x08 };
#pragma code-name (pop)
